clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1781,30 +1776,54 @@ private boolean processBackgroundEvents() {
-     * If the application thread simply blocked on the event after submitting it, the processing would deadlock.
-     * The logic herein is basically a loop that performs two tasks in each iteration:
-     *
-     * <ol>
-     *     <li>Process background events, if any</li>
-     *     <li><em>Briefly</em> wait for {@link CompletableApplicationEvent an event} to complete</li>
-     * </ol>
+     * When unsubscribing, the application thread enqueues an {@link UnsubscribeEvent} on the application event queue.
+     * That event will eventually trigger the rebalancing logic in the background thread.
+     * Critically, as part of this rebalancing work, any
+     * {@link ConsumerRebalanceListener#onPartitionsRevoked(Collection) rebalance listener callback} may need to be
+     * invoked.
      *
      * <p/>
      *
-     * Each iteration gives the application thread an opportunity to process background events, which may be
-     * necessary to complete the overall processing.
+     * There are a number of challenges that arise during this seemingly simple process:
      *
-     * <p/>
+     * <ul>
+     *     <li>
+     *         While most of the unsubscribe logic is performed on the background thread, the
+     *         {@link ConsumerRebalanceListener#onPartitionsRevoked(Collection) rebalance listener callback} must be
+     *         executed on the application thread. This requires a delicate dance between the two threads that is
+     *         orchestrated via the {@link ConsumerRebalanceListenerCallbackNeededEvent} and
+     *         {@link ConsumerRebalanceListenerCallbackCompletedEvent} events.
+     *     </li>
+     *     <li>
+     *         The user may or may not be using a callback handler. This can be deduced by
+     *         {@link SubscriptionState#rebalanceListener() checking for rebalance listener}.
+     *     </li>
+     *     <li>
+     *         If, at the time of unsubscribe, the consumer does not have any partitions assigned, the background
+     *         thread will <em>not</em> enqueue a {@link ConsumerRebalanceListenerCallbackNeededEvent} to signal the
+     *         application thread to execute the
+     *         {@link ConsumerRebalanceListener#onPartitionsRevoked(Collection) rebalance listener callback}. It's
+     *         technically possible to
+     *         {@link SubscriptionState#assignedPartitions() check if there are any partitions assigned} in the
+     *         application thread, but it's possible the assignment could change in the background thread. So the
+     *         application thread cannot blindly assume that a {@link ConsumerRebalanceListenerCallbackNeededEvent}
+     *         will appear in the background event queue even if a rebalance listener is in use.
+     *     </li>
+     *     <li>
+     *         The call to {@link ConsumerRebalanceListener#onPartitionsRevoked(Collection)} may take so long that
```

Comment:
`ConsumerRebalanceListener#onPartitionsRevoked` may be not executed if the timeout is too small, so could you please add this behavior to `close` method?

Reply:
> could you please add this behavior to `close` method?

My apologies—_which_ logic specifically do you want me to add to the `close()` method?